main.cc

```diff
@@ -1609,6 +1609,69 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            const qualified_name qualified_authenticator_name(auth::meta::AUTH_PACKAGE_NAME, cfg->authenticator());
+            const qualified_name qualified_role_manager_name(auth::meta::AUTH_PACKAGE_NAME, cfg->role_manager());
+
+            auth::service_config auth_config;
+            auth_config.authorizer_java_name = qualified_authorizer_name;
+            auth_config.authenticator_java_name = qualified_authenticator_name;
+            auth_config.role_manager_java_name = qualified_role_manager_name;
+
+            auth_service.start(std::move(perm_cache_config), std::ref(qp), std::ref(mm_notifier), std::ref(mm), auth_config).get();
+
+            auto start_auth_service = [&mm, maintanance_mode = cfg->maintanance_mode()] (std::any& stop_auth_service) {
+                auth_service.invoke_on_all(&auth::service::start, std::ref(mm), maintanance_mode).get();
+
+                stop_auth_service = defer_verbose_shutdown("auth service", [] {
+                    auth_service.stop().get();
+                });
+            };
+
+            api::set_server_authorization_cache(ctx, auth_service).get();
+            auto stop_authorization_cache_api = defer_verbose_shutdown("authorization cache api", [&ctx] {
+                api::unset_server_authorization_cache(ctx).get();
+            });
+
+            scheduling_group_key_config cql_sg_stats_cfg = make_scheduling_group_key_config<cql_transport::cql_sg_stats>();
+            cql_transport::controller cql_server_ctl(auth_service, mm_notifier, gossiper, qp, service_memory_limiter, sl_controller, lifecycle_notifier, *cfg, scheduling_group_key_create(cql_sg_stats_cfg).get0());
+            auto start_cql = [&cql_server_ctl, &dbcfg] (std::any& stop_cql) {
+                supervisor::notify("starting native transport");
+                with_scheduling_group(dbcfg.statement_scheduling_group, [&cql_server_ctl] {
+                    return cql_server_ctl.start_server();
+                }).get();
+
+                // FIXME -- this should be done via client hooks instead
+                stop_cql = defer_verbose_shutdown("native transport", [&cql_server_ctl] {
+                    cql_server_ctl.stop_server().get();
+                });
+            };
+
+            if (cfg->maintanance_mode()) {
+                startlog.info("entering maintenance mode.");
+
+                ss.local().start_maintanance_mode().get();
+
+                std::any stop_auth_service;
+                start_auth_service(stop_auth_service);
+
+                std::any stop_cql;
+                start_cql(stop_cql);
+
+                startlog.info("Scylla version {} initialization completed.", scylla_version());
+                stop_signal.wait().get();
+                startlog.info("Signal received; shutting down");
+                startlog.info("Scylla version {} shutdown complete.", scylla_version());
+                _exit(0);
+                return 0;
```

Comment:
I see we do not start the distributed system keyspaces. The maintenance mode should allow for manually fixing these too (allow writing to them).

Reply:
I didn't know that. Indeed, it makes sense to not start this then.